### ISSUE TITLE
[RFC] Builtin mean reduce function.

### ISSUE THREAD

**Issue by yzh119** [dmlc/dgl#785]
## 🚀 Feature
Many GNN models use mean reducer:
![image](https://user-images.githubusercontent.com/11773619/63561906-2f3c0000-c58e-11e9-9cb0-1846a00690ad.png)
Currently we only support sum/max/min/prod as builtin reducers, there should be a builtin mean reducer in `dgl.function.reducer`.

## Alternatives

Sum reduce + normalization(by in degrees) has the same the effect as mean reducer:
```python
import dgl.function as fn
g.update_all(fn.copy_u('h', 'm'), fn.sum('m', 'h'))
degs = graph.in_degrees().float().unsqueeze(-1)
g.ndata['h'] = g.ndata['h'] / degs
```
However, it's more convenient to call `fn.mean` directly.

## Pitch
Currently we do not have kernel support for mean reducers, I was wondering is it possible to decompose `fn.mean`to a reduce function `fn.sum` and an apply_node function that normalize the result.

Another concern is how do we deal with zero degree nodes? I prefer return an all-zero vector for these nodes.

**Comment by mufeili** [comment]
I agree with both points.

**Comment by jermainewang** [comment]
Since we don't have a fully functional int32 graph or CUDA graph, the question is where to generate the correct degree tensor. Apparently, we could not compute the degree tensor inside `binary_reduce` kernel function (https://github.com/dmlc/dgl/blob/master/python/dgl/kernel.py#L26) due to the aforementioned reason. I think as a workaround we could put the code right before the graph is converted to the required index type and device context. @ylfdq1118 could you provide some advises on this?

**Comment by ylfdq1118** [comment]
Personally, I think adding the norm at kernel function level is the cleanest way. But as Minjie mentioned, unfortunately, we only support graph structure query for 64-bit CPU graph.

In current scheduler design, builtin function interfaces are unified, and scheduler does not need to distinguish which reduce function is used (in other words, the code path for max/min/sum/prod is exactly the same). So it's actually not easy to add node degree for mean.

After stepping into this line:
https://github.com/dmlc/dgl/blob/708765f0a11d29b3d3f27857aeebfd10498f1937/python/dgl/runtime/scheduler.py#L869
you will lose access to the original 64-bit CPU DGLGraph object. So you need to check the reduce function and generate the node degree tensor before this line and pass it all the way down to kernel. And I believe this would require non-trivial and tedious code change.

Another easier but very dirty hack is still generating the degree tensor at kernel level. This is small modification to what you already did. For example, here:
https://github.com/dmlc/dgl/blob/fa3775c02fdf6601bf24afa4bc90d6e27c60c345/python/dgl/backend/mxnet/tensor.py#L431
before generating the degree tensor, you should first convert the graph back to a CPU 64-bit graph by `cpu_graph = self.graph.copy_to(cpu_ctx).asbits(64)`. After generating the node degree tensor, you can then copy it to device and convert to the same integer type as `self.graph`.

**Comment by yzh119** [comment]
@ylfdq1118 @jermainewang to minimize changes to the codebase. I decide to compute the in degree tensor using another kernel call. Please refer to #787 .

**Comment by jermainewang** [comment]
Takeaway from this RFC. I wonder what's the best way to implement this in the future? We can assume that the graph index can generate degree tensor easily regardless of its device context and integer type. It looks like we could modify the executor code:
https://github.com/dmlc/dgl/blob/708765f0a11d29b3d3f27857aeebfd10498f1937/python/dgl/runtime/ir/executor.py#L1062-L1078

1. Use the graph created in L1066 to get degree tensor.
2. Divide `self.ret.data` by the degree tensor after L1076.

Since the operators involved here (i.e, `F.binary_reduce` and division) are recorded by autograd system, the backward computation should be correct.

@ylfdq1118 any comments?

**Comment by ylfdq1118** [comment]
I think @jermainewang 's suggestion is nice and clean.
